addons/account/models/account_invoice.py

```diff
@@ -1313,13 +1313,18 @@ def _refund_cleanup_lines(self, lines):
             :return: list of command tuple for one2many line creation [(0, 0, dict of valueis), ...]
         """
         result = []
+        model_invoice_line = lines._name == 'account.invoice.line'
         for line in lines:
             values = {}
             for name, field in line._fields.items():
                 if name in MAGIC_COLUMNS:
                     continue
                 elif field.type == 'many2one':
-                    values[name] = line[name].id
+                    if model_invoice_line and line.invoice_id.type == 'out_invoice' and field.name == 'account_id':
```

Comment:
@oco I didn't get the part
`
"for name, field in line._fields.items():" would then stay like it was before, except it would now be included in the else part of the condition. This would avoid useless iterations and would be nicer anyway.
`
In my view, we can and have to check `account_id` in `elif field.type == 'many2one';` block only since we want to set different value only for `account_id`(an m2o field) and that too when `lines` is of `account.invoice.line` recordset and not when `account.invoice.tax` recordset. Can you please show how it can be done ?

rest of the improvements are done.

Thanks.

Reply:
You're right, I had missed a little subtlety of this piece of code. You may ignore this comment ;)